### PR TITLE
Improve reactive server span name generation

### PR DESCRIPTION
**Pull request by lightoze**
The patch implements `HttpAdapter.methodFromResponse(response)` and `HttpAdapter.route(response)` which are required for Brave HTTP span name generation (`HttpParser.spanNameFromRoute`).

### CODE REVIEW

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java [hunk 1]
```diff
@@ -156,6 +158,11 @@ TraceKeys traceKeys() {
 								addClassMethodTag(handlerMethod, span);
 								addClassNameTag(handlerMethod, span);
 							}
+							Object httpRoute = exchange
+									.getAttribute(HandlerMapping.BEST_MATCHING_PATTERN_ATTRIBUTE);
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java]: the contract is empty string when route is available (which I think is always the case here), just no match.

Ex. in webmvc we do..
```
    Object httpRoute = request.getAttribute(BEST_MATCHING_PATTERN_ATTRIBUTE);
    request.setAttribute("http.route", httpRoute != null ? httpRoute.toString() : "");
```

So here probably

```
    Object httpRoute = request.getAttribute(BEST_MATCHING_PATTERN_ATTRIBUTE);
    response.httpRoute = httpRoute != null ? httpRoute.toString() : "";
```

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java [hunk 2]
```diff
@@ -156,6 +158,11 @@ TraceKeys traceKeys() {
...
+							if (httpRoute != null) {
+								response.httpRoute = httpRoute.toString();
+							}
 							addResponseTagsForSpanWithoutParent(exchange, response, span);
 							handler().handleSend(response, t, span);
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java]: I'd move the new DecoratedServerHttpResponse(exchange.getResponse()) here. it is only used by handleSend

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java [hunk 3]
```diff
@@ -255,6 +262,15 @@ private void addClassNameTag(Object handler, Span span) {
 		return ORDER;
 	}
 
+	static final class DecoratedServerHttpResponse extends ServerHttpResponseDecorator {
+
+		String method, httpRoute;
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java]: if you move instantiating this to where handleSend is called, you can make these final and always assign a value

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java (hunk 2 above)
Comment (lightoze) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java]: Valid point. Speaking of that - is there any reason for `request` and `response` being referenced from lambdas instead of retrieving them from the exchange? Not a big optimization but still...

Comment (adriancole) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java]: you can fix this so long as other tests still pass. We haven't optimized this code yet for performance.

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java (hunk 1 above)
Comment (lightoze) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java]: There you already know for sure that request is handled by MVC. In this case it is not always true (think reactive functions for example). Also I was not able to reproduce null `BEST_MATCHING_PATTERN_ATTRIBUTE` while staying in MVC context. However I noticed following corner cases:
- default error page provides `/**` as a pattern
- for catch-all `GetMapping` provided pattern equals to URI - whatever the client supplies

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java [hunk 4]
```diff
@@ -192,6 +196,15 @@ TraceKeys traceKeys() {
 						}));
 	}
 
+	private static String getHttpRoute(ServerWebExchange exchange) {
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java]: maybe make the notes about the corner cases as comments here (better as unit tests, but if no tests, then at least comments keeps the knowledge around)

I suspect that if you are in the `attribute instanceof HandlerMethod` block, then you'd know you are in a controller, so if the pattern string is null in this case, you'd coerse to empty to fit the contract.

Comment (lightoze) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java]: I am still convinced that coercion is not needed because pattern attribute is never null for MVC calls. For other calls returning empty string will overwrite span name even if user have already manually set custom one - can be the case with router functions.

I'll add some comments, extend existing test to check generated span name and let's call it a day, OK? :)

Comment (adriancole) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java]: to see what I mean, you can clone brave locally. add a test that makes the thing you are concerned about, for example, in webmvc. the current approach to span customization (for example if you have a better name) is to have your own parser.

https://github.com/openzipkin/brave/blob/master/instrumentation/http-tests/src/main/java/brave/test/http/ITHttpServer.java

Even if we dodge it here, if there is a problem, the trap is set elsewhere. Not trying to be arbitrary, just this sort of inconsistency limits the ability of a couple people to maintain projects that a lot of people use.

Comment (lightoze) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java]: About `HttpAdapter.route` result - I've read the docs again and it says "Null indicates this instrumentation doesn't understand http routes". `TraceWebFilter` does support routes so its implementation of HttpAdapter should never return null. Is that correct?

I feel like I've lost the direction of your though. What do you suggest?

Comment (adriancole) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java]: At least when "attribute instanceof HandlerMethod" I suspect we can coerce null to empty. I think making sure that this happens (if it is correct that attribute instanceof HandlerMethod when defined on an annotated class), then we make progress and keep things consistent. So, when you are inside the  "attribute instanceof HandlerMethod" if statement make sure getHttpRoute null coerces to empty somehow

There may be other cases, too, but we can defer that to later tests.

Make sense?

Independently, we need to put a subclass of ITHttpServer here, because it is better to show not tell. This is something I will try to do offline.

Comment (lightoze) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java]: OK, I redid the whole patch according to this. Also added a test for router function to demonstrate span naming behaviour.

**Comment by adriancole** [comment]
thanks for the help!

**Comment by marcingrzejszczak** [comment]
Great work!
